### ISSUE TITLE
Calendar: current date is not being highlighted.

### ISSUE THREAD

**Issue by sprnza** [copycat-killer/lain#300]
Hi there!
I can't remember exactly the update I've started facing this issue but calendar's stopped highlighting the current date.
I use this code to attach the widget:

```
cal = lain.widgets.calendar({attach_to = {mytextclock}})
```

In a terminal `cal` returns the date highlighted properly.

Just tried a default awesom's config- it behaves the same way- no current date highlighting.

**Comment by copycat-killer** [comment]
Try [this](https://github.com/copycat-killer/lain/commit/32eddd514f8ccc35eeaa128a441f6a47708e6a04).

**Comment by sprnza** [comment]
It's already there. I've updated today morning.

**Comment by dummys** [comment]
Why closing the issue ? It's still not working right ?

**Comment by sprnza** [comment]
@dummys it's another issue was closed. This one is not.

**Comment by copycat-killer** [comment]
> It's already there. I've updated today morning.

Sorry, I mean try reverting to the previous code.

**Comment by sprnza** [comment]
@copycat-killer The previous code doesn't work either

**Comment by copycat-killer** [comment]
Is `/usr/bin/cal --color=always` the right command to have a colorized `cal` input in your distro?

For instance, if you're using Ubuntu, it's `/usr/bin/cal -h` instead.

Or did you do specific `/etc/terminal-colors.d` settings?

**Comment by sprnza** [comment]
I use ArchLinux so `--color=always` is the option I should use. Either `cal` and `cal --color=always` returns the current date highlighted properly in terminal but nothing works properly in the calendar's notification.

**Comment by copycat-killer** [comment]
I'm on Arch too. This is what I use:

```lua
lain.widgets.calendar {
    cal = "/usr/bin/cal --color=always",
    attach_to = { mytextclock }
}
```

and it works (simple `cal` doesn't).

You can try adjusting the matching pattern in Lua:

```lua
$ lua
> cal = io.popen("/usr/bin/cal --color=always"):read("*a")
> cal:gsub("%c%[%d+[m]?%d+%c%[%d+[m]?", "DID I PICK IT?")
```

If you read the blurb, then it's working.

Of course, I'll work on a solution myself, but in the meantime you can test.

**Comment by sprnza** [comment]
It's working in lua, but not in Awesome, the calendar doesn't have the highlighting. I just tried a default config(with default theme) with your snippet added. It's not working.

**Comment by dummys** [comment]
Not working in my setup too.
Arch linux as well.

**Comment by 2009** [comment]
This is broken for me now too even after adding ```cal = "/usr/bin/cal --color=always```.
It looks like pattern didn't match my color codes that cal was producing.
```
    February 2017   
Su Mo Tu We Th Fr Sa
         ^[[7m 1^[[27m  2  3  4 
 5  6  7  8  9 10 11 
12 13 14 15 16 17 18 
19 20 21 22 23 24 25 
26 27 28  
```
I also had a problem with the icon not showing up because os.date was returning 01 instead of 1.
I created a pull request of what I did locally.

I think it would probably be better to change the pattern so it matches the output of the cal command without the --color=always option.

**Comment by sprnza** [comment]
Sorry for misleading comment:

> $ lua
> cal = io.popen("/usr/bin/cal --color=always"):read("*a")
> cal:gsub("%c%[%d+[m]?%d+%c%[%d+[m]?", "DID I PICK IT?")

It's not being picked.

**Comment by sprnza** [comment]
```
cal:gsub("%c%[%d+[m]?%d+%c%[%d+[m]?", "DID I PICK IT?")
```
doesn't work either
```
cal --color=always|less

    February 2017   
Su Mo Tu We Th Fr Sa
         ESC[7m 1ESC[27m  2  3  4 
 5  6  7  8  9 10 11 
12 13 14 15 16 17 18 
19 20 21 22 23 24 25 
26 27 28

**Comment by mbuelte** [comment]
It might be the locale settings which leads into showing the day of month without the leading zero. Therefor we also make sure that the "today" variable exclude it with:

`local today = os.date("%e")`

And then markup the right one:

`ws = ws:gsub(today.." ", markup.bold(markup.color(bg, fg, today)))`

The regex for the global substitution needs to be fixed ;-)

**Comment by copycat-killer** [comment]
@mbuelte You beat me on time. ;-)

**Comment by mbuelte** [comment]
@copycat-killer, I'm afraid the regex still doensn't work (for me). Strangely enough that it match from the lua shell, but only if I append '--color=always' - it seems you've forgotten that in the latest commit.

**Comment by sprnza** [comment]
Doesn't for me either.

**Comment by copycat-killer** [comment]
And of course it's working only for me.

Does pull #303 works for you?

**Comment by mbuelte** [comment]
Unfortunately not

**Comment by mbuelte** [comment]
How about such a solution? This may not be exactly nice, but it works for me:

```
...
local today = os.date("%e")
...
if(today:match("^%s%d")) then
  ws = ws:gsub(today.." ", markup.bold(markup.color(bg, fg, today)))
else
  ws = ws:gsub(today, markup.bold(markup.color(bg, fg, today)))
end
...
```

**Comment by copycat-killer** [comment]
Keep `local today = os.date("%e")` and try this:

```diff
-ws = ws:gsub("%c%[+%d+[m]?%s?%d+%c%[+%d+[m]?", markup.bold(markup.color(bg, fg, today)))
+ws = ws:gsub("%c%[%d+[m]?" .. today .. "%c%[%d+[m]?", markup.bold(markup.color(bg, fg, today)))
```

**Comment by mbuelte** [comment]
Sorry, same problem, the regex won't match here :(

**Comment by copycat-killer** [comment]
Since it worked for you in the shell, are you sure you don't have an environment problem?

Try setting `env` in your `cal` command (or something related), and also set `awful.util.shell` to your shell, in your `rc.lua`, before calling `lain.widgets.cal`.

**Comment by mbuelte** [comment]
That's driving me nuts. I don't think there is an issue with the environment, everything looks similiar. I wonder, why the escape sequences doesn't came up in the output. Afterwards I wrote a little bash script that simulate the output of `cal`:

```
echo -e " Su Mo Tu We Th Fr Sa\n" \
  "         \e[7m 1\e[27m  2  3  4\n" \
  " 5  6  7  8  9 10 11\n"
```
And I see the escape sequences in the output, your regex matches and is highlighted as wished.

**Comment by copycat-killer** [comment]
I don't know. Did you try setting `awful.util.shell`?

Try to understand why [this](https://github.com/copycat-killer/lain/issues/300#issuecomment-276707194) works on your system. If you can't come up with a solution... you can try to write the above one better.

**Comment by sprnza** [comment]
> Edit: This works even if cal was run without --color=always.
> @sprnza can you confirm this?

Yep, the only working variant is one you suggested. 

> EDIT:

 You can use naughty.notify{text=ws, timeout=0} before ws:gsub(...) to print the pre-processing text.

Actually I used this in helpers.lua inside the async function. It returns `/usr/bin/cal --color=always` for cmd and there is no the escape sequences in stdout.

**Comment by mbuelte** [comment]
@sprnza That's the same I noticed here. When you try my bash script instead of `cal` it will show the escape sequences in the output. It seems somewhat ignores this kind of escape characters and others not.

**Comment by sprnza** [comment]
@mbuelte Yep, your script is being processed just fine

**Comment by copycat-killer** [comment]
There's a problem with @mbuelte [solution](https://github.com/copycat-killer/lain/issues/300#issuecomment-276763994): it will highlight every `today` match. So, for instance, if you have `cal -w`, or any other instance of `os.date("%e")` in you output, it will be highlighted.

Can you try setting `cal = "bash -c cal"` or something similar?

**Comment by sprnza** [comment]
> Can you try setting cal = "bash -c cal" or something similar?
Doesn't work.

What I've found that if I pipe cal to a file, then convert it to a scipt and call it. It's being displayed properly. So it's cal who is misbehaving. It doesn't return escape sequences under awesome env.
```
#!/bin/bash
echo "    February 2017   
Su Mo Tu We Th Fr Sa
         ^[[7m 1^[[27m  2  3  4 
 5  6  7  8  9 10 11 
12 13 14 15 16 17 18 
19 20 21 22 23 24 25 
26 27 28            " 
```

**Comment by sprnza** [comment]
Even if I use
```
cal_output = io.popen("LANG=C /usr/bin/cal --color=always"):read("*a")
```
inside rc.lua (wich is working in lua shell) and then return the variable using awesome-client it doesn't have the escape sequences.

**Comment by sprnza** [comment]
`awful.util.shell = "/usr/bin/bash"` doesn't help

**Comment by mbuelte** [comment]
Same here, even such things like `cal --color=always | cat -v` which shows the escape sequences in normal env, aren't shown within naughty.notify.

**Comment by sprnza** [comment]
Works perfectly under Xephyr.

**Comment by sprnza** [comment]
Works perfectly if I start awesomewm using startx + xinitrc

**Comment by copycat-killer** [comment]
@sprnza What works perfectly? The standard `lain.widgets.calendar` or some patched version posted early?

How were you starting awesome before?

**Comment by sprnza** [comment]
The standard `lain.widgets.calendar` works like a rock from startx and doesn't using lightdm.

**Comment by copycat-killer** [comment]
@mbuelte How do you start awesome?

**Comment by sprnza** [comment]
adding `TERM=linux` in /etc/environment solves the "issue"

**Comment by mbuelte** [comment]
Great! I'll try it out soon. I start awesome by xdg, but using startx doesn't solved the issue.

**Comment by copycat-killer** [comment]
@sprnza Can you try with `cal = "TERM=linux cal --color=always" or  `cal = "/usr/bin/env TERM=linux cal --color=always" instead of modifying environment files?

**Comment by sprnza** [comment]
> `cal = "/usr/bin/env TERM=linux cal --color=always"`
this one works.

**Comment by sprnza** [comment]
@copycat-killer sorry for had being disturbing you with this unrelated issue. And thank you!

**Comment by copycat-killer** [comment]
@mbuelte Let me know if it works for you too.

**Comment by mbuelte** [comment]
@copycat-killer @sprnza Thanks a lot!

`cal = "/usr/bin/env /usr/bin/cal --color=always"`

Seems to be enough, it doesn't rely on the TERM variable - nevermind. I won't think about that anymore :)

**Comment by dummys** [comment]
cal = "/usr/bin/env TERM=linux /usr/bin/cal --color=always"

This one works for me too !
thanks

**Comment by copycat-killer** [comment]
@mbuelte @sprnza @dummys 

Is there a problem if I revert this [change](https://github.com/copycat-killer/lain/issues/300#issuecomment-276723693)? It's not working for me.

**Comment by sprnza** [comment]
No problems with the reverting.

**Comment by mbuelte** [comment]
No problem, thanks!

**Comment by sebras** [comment]
I'll quietly add this here to help the next person. It might be connected to your issue, but if not, don't mind me. :)

I encountered the issue of `cal` in the terminal not highlighting today's date on one of my computer, but *did* on another one. Which searching for a solution I came across this discussion. For me your suggestion `--color=always` is viable since it is not an option that exists in `cal` on Debian. Continuing searching revealed [a similar discussion](https://askubuntu.com/questions/1164415/cal-h-does-not-produce-expected-output), which led me to [this bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=904839) that was fixed by [this commit](https://salsa.debian.org/meskes/bsdmainutils/-/commit/640967a51c9623362beab94ba6e87f74a9448ca0) in the ncal package in Debian. The solution for me was to put `alias cal="if [ -t 1 ] ; then ncal -b ; else /usr/bin/cal ; fi"` in my `.bashrc` which gives expected output for scripts, e.g. `cal | cat` and expected output for humans `cal`.
